src/server/game/Entities/Unit/Unit.cpp

```diff
@@ -467,6 +467,7 @@ void Unit::Update(uint32 p_time)
 
     UpdateSplineMovement(p_time);
     i_motionMaster->UpdateMotion(p_time);
+    UpdateUnderwaterState(GetMap(), GetPositionX(), GetPositionY(), GetPositionZ());
```

Comment:
I removed it from Unit::UpdatePosition and moved it in Unit::Update. This method better belong here than in the MotionMaster.